README.md
#Sound Synthesizer

[![Build Status](https://travis-ci.org/LucianBuzzo/sound-synthesizer.svg?branch=master)](https://travis-ci.org/LucianBuzzo/sound-synthesizer)

A sound synthesizer written in Javascript.

Based on https://github.com/cwilso/midi-synth


CHANGELOG.md
## v0.0.2 - 2018-11-23

* Add travis CI configuration #1 #2 [Lucian]

## v0.0.3 - 2018-11-23

* Add correct encrypted npm API key to travis config [Lucian]


lib/voice.js
const frequencyFromNoteNumber = function (note) {
  return 440 * Math.pow(2, (note - 69) / 12)
}

const filterFrequencyFromCutoff = function (pitch, cutoff, ctx) {
  var nyquist = 0.5 * ctx.sampleRate

  //    var filterFrequency = Math.pow(2, (9 * cutoff) - 1) * pitch
  var filterFrequency = Math.pow(2, 9 * cutoff - 1) * pitch
  if (filterFrequency > nyquist) {
    filterFrequency = nyquist
  }
  return filterFrequency
}

const Voice = function Voice (note, velocity, synth) {
  this.originalFrequency = frequencyFromNoteNumber(note)
  this.currentOsc2Octave = synth.currentOsc2Octave
  this.currentOsc2Detune = synth.currentOsc2Detune
  this.currentOsc1Octave = synth.currentOsc1Octave
  this.currentOsc1Detune = synth.currentOsc1Detune
  this.currentPitchWheel = synth.currentPitchWheel
  this.currentEnvR = synth.currentEnvR
  this.currentFilterEnvR = synth.currentFilterEnvR
  this.currentFilterCutoff = synth.currentFilterCutoff
  this.currentFilterMod = synth.currentFilterMod
  let {
    audioContext,
    currentEnvA,
    currentEnvD,
    currentEnvS,
    currentFilterEnv,
    currentFilterEnvA,
    currentFilterEnvD,
    currentFilterEnvS,
    currentFilterQ,
    currentModFrequency,
    currentModOsc1,
    currentModOsc2,
    currentModWaveform,
    currentOsc1Mix,
    currentOsc1Waveform,
    currentOsc2Mix,
    currentOsc2Waveform,
    effectChain,
    modOscFreqMultiplier,
    waveforms
  } = synth

  this.audioContext = audioContext

  // create osc 1
  this.osc1 = audioContext.createOscillator()
  this.updateOsc1Frequency()
  this.osc1.type = currentOsc1Waveform

  this.osc1Gain = audioContext.createGain()
  this.osc1Gain.gain.value = 0.005 * currentOsc1Mix
  this.osc1.connect(this.osc1Gain)

  // create osc 2
  this.osc2 = audioContext.createOscillator()
  this.updateOsc2Frequency()
  this.osc2.type = waveforms[currentOsc2Waveform]

  this.osc2Gain = audioContext.createGain()
  this.osc2Gain.gain.value = 0.005 * currentOsc2Mix
  this.osc2.connect(this.osc2Gain)

  // create modulator osc
  this.modOsc = audioContext.createOscillator()
  this.modOsc.type = currentModWaveform
  this.modOsc.frequency.value = currentModFrequency * modOscFreqMultiplier

  this.modOsc1Gain = audioContext.createGain()
  this.modOsc.connect(this.modOsc1Gain)
  this.modOsc1Gain.gain.value = currentModOsc1 / 10
  this.modOsc1Gain.connect(this.osc1.frequency) // tremolo

  this.modOsc2Gain = audioContext.createGain()
  this.modOsc.connect(this.modOsc2Gain)
  this.modOsc2Gain.gain.value = currentModOsc2 / 10
  this.modOsc2Gain.connect(this.osc2.frequency) // tremolo

  // create the LP filter
  this.filter1 = audioContext.createBiquadFilter()
  this.filter1.type = 'lowpass'
  this.filter1.Q.value = currentFilterQ
  this.filter1.frequency.value = Math.pow(2, this.currentFilterCutoff)
  this.filter2 = audioContext.createBiquadFilter()
  this.filter2.type = 'lowpass'
  this.filter2.Q.value = currentFilterQ
  this.filter2.frequency.value = Math.pow(2, this.currentFilterCutoff)

  this.osc1Gain.connect(this.filter1)
  this.osc2Gain.connect(this.filter1)
  this.filter1.connect(this.filter2)

  // connect the modulator to the filters
  this.modFilterGain = audioContext.createGain()
  this.modOsc.connect(this.modFilterGain)
  this.modFilterGain.gain.value = this.currentFilterMod * 24
  this.modFilterGain.connect(this.filter1.detune) // filter tremolo
  this.modFilterGain.connect(this.filter2.detune) // filter tremolo

  // create the volume envelope
  this.envelope = audioContext.createGain()
  this.filter2.connect(this.envelope)
  this.envelope.connect(effectChain)

  // set up the volume and filter envelopes
  var now = audioContext.currentTime
  var envAttackEnd = now + currentEnvA / 20.0

  this.envelope.gain.value = 0.0
  this.envelope.gain.setValueAtTime(0.0, now)
  this.envelope.gain.linearRampToValueAtTime(1.0, envAttackEnd)
  this.envelope.gain.setTargetAtTime(currentEnvS / 100.0, envAttackEnd, currentEnvD / 100.0 + 0.001)

  var filterAttackLevel = currentFilterEnv * 72 // Range: 0-7200: 6-octave range
  var filterSustainLevel = filterAttackLevel * currentFilterEnvS / 100.0 // range: 0-7200
  var filterAttackEnd = currentFilterEnvA / 20.0

  if (!filterAttackEnd) {
    filterAttackEnd = 0.05 // tweak to get target decay to work properly
  }
  this.filter1.detune.setValueAtTime(0, now)
  this.filter1.detune.linearRampToValueAtTime(filterAttackLevel, now + filterAttackEnd)
  this.filter2.detune.setValueAtTime(0, now)
  this.filter2.detune.linearRampToValueAtTime(filterAttackLevel, now + filterAttackEnd)
  this.filter1.detune.setTargetAtTime(filterSustainLevel, now + filterAttackEnd, currentFilterEnvD / 100.0)
  this.filter2.detune.setTargetAtTime(filterSustainLevel, now + filterAttackEnd, currentFilterEnvD / 100.0)

  this.osc1.start(0)
  this.osc2.start(0)
  this.modOsc.start(0)
}

Voice.prototype.setModWaveform = function (value) {
  this.modOsc.type = value
}

Voice.prototype.updateModFrequency = function (value) {
  this.modOsc.frequency.value = value
}

Voice.prototype.updateModOsc1 = function (value) {
  this.modOsc1Gain.gain.value = value / 10
}

Voice.prototype.updateModOsc2 = function (value) {
  this.modOsc2Gain.gain.value = value / 10
}

Voice.prototype.setOsc1Waveform = function (value) {
  this.osc1.type = value
}

Voice.prototype.updateOsc1Frequency = function () {
  // -2 because osc1 is 32', 16', 8'
  let val = this.originalFrequency * Math.pow(2, this.currentOsc1Octave - 2)
  this.osc1.frequency.value = val
  this.osc1.detune.value = this.currentOsc1Detune + this.currentPitchWheel * 500 // value in cents - detune major fifth.
}

Voice.prototype.updateOsc1Mix = function (value) {
  this.osc1Gain.gain.value = 0.005 * value
}

Voice.prototype.setOsc2Waveform = function (value) {
  this.osc2.type = value
}

Voice.prototype.updateOsc2Frequency = function () {
  this.osc2.frequency.value = this.originalFrequency * Math.pow(2, this.currentOsc2Octave - 1)
  this.osc2.detune.value = this.currentOsc2Detune + this.currentPitchWheel * 500 // value in cents - detune major fifth.
}

Voice.prototype.updateOsc2Mix = function (value) {
  this.osc2Gain.gain.value = 0.005 * value
}

Voice.prototype.setFilterCutoff = function (value) {
  var filterFrequency = Math.pow(2, value)
  this.filter1.frequency.value = filterFrequency
  this.filter2.frequency.value = filterFrequency
}

Voice.prototype.setFilterQ = function (value) {
  this.filter1.Q.value = value
  this.filter2.Q.value = value
}

Voice.prototype.setFilterMod = function () {
  this.modFilterGain.gain.value = this.currentFilterMod * 24
}

Voice.prototype.noteOff = function () {
  var now = this.audioContext.currentTime
  var release = now + this.currentEnvR / 10.0
  filterFrequencyFromCutoff(this.originalFrequency, this.currentFilterCutoff / 100 * (1.0 - this.currentFilterEnv / 100.0), this.audioContext)

  this.envelope.gain.cancelScheduledValues(now)
  this.envelope.gain.setValueAtTime(this.envelope.gain.value, now) // this is necessary because of the linear ramp
  this.envelope.gain.setTargetAtTime(0.0, now, this.currentEnvR / 100)
  this.filter1.detune.cancelScheduledValues(now)
  this.filter1.detune.setTargetAtTime(0, now, this.currentFilterEnvR / 100.0)
  this.filter2.detune.cancelScheduledValues(now)
  this.filter2.detune.setTargetAtTime(0, now, this.currentFilterEnvR / 100.0)

  this.osc1.stop(release)
  this.osc2.stop(release)
}

module.exports = Voice


lib/waveshaper.js
const threshold = -27 // dB
const headroom = 21 // dB

function e4 (x, k) {
  return 1.0 - Math.exp(-k * x)
}

function dBToLinear (db) {
  return Math.pow(10.0, 0.05 * db)
}

function shape (x) {
  var linearThreshold = dBToLinear(threshold)
  var linearHeadroom = dBToLinear(headroom)

  var maximum = 1.05 * linearHeadroom * linearThreshold
  var kk = (maximum - linearThreshold)

  var sign = x < 0 ? -1 : +1
  var absx = Math.abs(x)

  var shapedInput = absx < linearThreshold ? absx : linearThreshold + kk * e4(absx - linearThreshold, 1.0 / kk)
  shapedInput *= sign

  return shapedInput
}

function generateColortouchCurve (curve) {
  var n = 65536
  var n2 = n / 2

  for (var i = 0; i < n2; ++i) {
    let x = i / n2
    x = shape(x)

    curve[n2 + i] = x
    curve[n2 - i - 1] = -x
  }

  return curve
}

function WaveShaper (context) {
  this.context = context
  var waveshaper = context.createWaveShaper()
  var preGain = context.createGain()
  var postGain = context.createGain()
  preGain.connect(waveshaper)
  waveshaper.connect(postGain)
  this.input = preGain
  this.output = postGain
  this.waveshaper = waveshaper

  var curve = new Float32Array(65536) // FIXME: share across instances
  generateColortouchCurve(curve)
  waveshaper.curve = curve
}

WaveShaper.prototype.setDrive = function (drive) {
  this.input.gain.value = drive
  var postDrive = Math.pow(1 / drive, 0.6)
  this.output.gain.value = postDrive
}

module.exports = WaveShaper


lib/synthesizer.js
/* global AudioContext, XMLHttpRequest */
/**
 *
 * Borrowed from https://github.com/cwilso/midi-synth
 */

const WaveShaper = require('./waveshaper')
const Voice = require('./voice')

const x = { foo: 'bar' }

const { foo } = x

console.log(foo)

class Synthesizer {
  constructor (audioContext = new AudioContext()) {
    this.audioContext = audioContext

    this.voices = []

    this.currentOctave = 3
    this.modOscFreqMultiplier = 1
    this.moDouble = false
    this.moQuadruple = false

    this.currentPitchWheel = 0.0

    this.waveforms = ['sine', 'square', 'sawtooth', 'triangle']

    // This is the "initial patch"
    this.currentModWaveform = 'sine'
    this.currentModFrequency = 2.1 // Hz * 10 = 2.1
    this.currentModOsc1 = 15
    this.currentModOsc2 = 17

    this.currentOsc1Waveform = 'sawtooth'
    this.currentOsc1Octave = 0 // 32'
    this.currentOsc1Detune = 0 // 0
    this.currentOsc1Mix = 50.0 // 50%

    this.currentOsc2Waveform = 2 // SAW
    this.currentOsc2Octave = 0 // 16'
    this.currentOsc2Detune = -25 // fat detune makes pretty analogue-y sound.  :)
    this.currentOsc2Mix = 50.0 // 0%

    this.currentFilterCutoff = 8
    this.currentFilterQ = 7.0
    this.currentFilterMod = 21
    this.currentFilterEnv = 56

    this.currentEnvA = 2
    this.currentEnvD = 15
    this.currentEnvS = 68
    this.currentEnvR = 5

    this.currentFilterEnvA = 5
    this.currentFilterEnvD = 6
    this.currentFilterEnvS = 5
    this.currentFilterEnvR = 7

    this.currentDrive = 38
    this.currentRev = 32
    this.currentVol = 75
    // end initial patch

    let keys = new Array(256)
    // Lower row: zsxdcvgbhnjm...
    keys[16] = 41 // = F2
    keys[65] = 42
    keys[90] = 43
    keys[83] = 44
    keys[88] = 45
    keys[68] = 46
    keys[67] = 47
    keys[86] = 48 // = C3
    keys[71] = 49
    keys[66] = 50
    keys[72] = 51
    keys[78] = 52
    keys[77] = 53 // = F3
    keys[75] = 54
    keys[188] = 55
    keys[76] = 56
    keys[190] = 57
    keys[186] = 58
    keys[191] = 59

    // Upper row: q2w3er5t6y7u...
    keys[81] = 60 // = C4 ("middle C")
    keys[50] = 61
    keys[87] = 62
    keys[51] = 63
    keys[69] = 64
    keys[82] = 65 // = F4
    keys[53] = 66
    keys[84] = 67
    keys[54] = 68
    keys[89] = 69
    keys[55] = 70
    keys[85] = 71
    keys[73] = 72 // = C5
    keys[57] = 73
    keys[79] = 74
    keys[48] = 75
    keys[80] = 76
    keys[219] = 77 // = F5
    keys[187] = 78
    keys[221] = 79
    keys[220] = 80

    this.keys = keys

    this.effectChain = null
    this.waveshaper = null
    this.volNode = null
    this.revNode = null
    this.revGain = null
    this.revBypassGain = null
    this.compressor = null

    this.initAudio(audioContext)
  }

  noteOn (note, velocity) {
    if (!this.voices[note]) {
      // Create a new synth node
      this.voices[note] = new Voice(note, velocity, this)
    }
  }

  noteOff (note) {
    if (this.voices[note]) {
      // Shut off the note playing and clear it
      this.voices[note].noteOff()
      this.voices[note] = null
    }
  }

  polyPressure (noteNumber, value) {
    if (this.voices[noteNumber]) {
      this.voices[noteNumber].setFilterQ(value * 20)
    }
  }

  updateModWaveform (value) {
    if (this.waveforms.indexOf(value) === -1) {
      throw new RangeError()
    }
    this.currentModWaveform = value
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].setModWaveform(value)
      }
    }
  }

  updateModFrequency (value) {
    if (value < 0 || value > 100) {
      throw new RangeError('value must be between 0 and 100')
    }
    value /= 10
    this.currentModFrequency = value
    var oscFreq = this.currentModFrequency * this.modOscFreqMultiplier
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].updateModFrequency(oscFreq)
      }
    }
  }

  updateModOsc1 (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentModOsc1 = value
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].updateModOsc1(this.currentModOsc1)
      }
    }
  }

  updateModOsc2 (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentModOsc2 = value
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].updateModOsc2(this.currentModOsc2)
      }
    }
  }

  // Range is in HZ
  updateFilterCutoff (value) {
    if (value < 20 || value > 20000) {
      throw new RangeError()
    }
    value = Math.sqrt(value) / 10
    this.currentFilterCutoff = value
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].setFilterCutoff(value)
      }
    }
  }

  updateFilterQ (value) {
    if (value < 0 || value > 20) {
      throw new RangeError()
    }
    this.currentFilterQ = value
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].setFilterQ(value)
      }
    }
  }

  updateFilterMod (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentFilterMod = value
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].setFilterMod(value)
      }
    }
  }

  updateFilterEnv (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentFilterEnv = value
  }

  updateOsc1Wave (value) {
    if (this.waveforms.indexOf(value) === -1) {
      throw new RangeError()
    }
    this.currentOsc1Waveform = value
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].setOsc1Waveform(value)
      }
    }
  }

  updateOsc1Octave (value) {
    let options = ["32'", "16'", "8'"]
    console.log(value)
    if (options.indexOf(value) === -1) {
      throw new RangeError()
    }
    this.currentOsc1Octave = options.indexOf(value)
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].updateOsc1Frequency()
      }
    }
  }

  updateOsc1Detune (value) {
    if (value < -1200 || value > 1200) {
      throw new RangeError()
    }
    this.currentOsc1Detune = value
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].updateOsc1Frequency()
      }
    }
  }

  updateOsc1Mix (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentOsc1Mix = value
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].updateOsc1Mix(value)
      }
    }
  }

  updateOsc2Wave (value) {
    if (this.waveforms.indexOf(value) === -1) {
      throw new RangeError()
    }
    this.currentOsc2Waveform = value
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].setOsc2Waveform(this.waveforms[this.currentOsc2Waveform])
      }
    }
  }

  updateOsc2Octave (value) {
    let options = ["16'", "8'", "4'"]
    if (options.indexOf(value) === -1) {
      throw new RangeError()
    }
    this.currentOsc2Octave = value
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].updateOsc2Frequency()
      }
    }
  }

  updateOsc2Detune (value) {
    if (value < -1200 || value > 1200) {
      throw new RangeError()
    }
    this.currentOsc2Detune = value
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].updateOsc2Frequency()
      }
    }
  }

  updateOsc2Mix (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentOsc2Mix = value
    for (var i = 0; i < 255; i++) {
      if (this.voices[i]) {
        this.voices[i].updateOsc2Mix(value)
      }
    }
  }

  updateEnvA (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentEnvA = value
  }

  updateEnvD (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentEnvD = value
  }

  updateEnvS (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentEnvS = value
  }

  updateEnvR (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentEnvR = value
  }

  updateFilterEnvA (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentFilterEnvA = value
  }

  updateFilterEnvD (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentFilterEnvD = value
  }

  updateFilterEnvS (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentFilterEnvS = value
  }

  updateFilterEnvR (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentFilterEnvR = value
  }

  updateDrive (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.currentDrive = value
    this.waveshaper.setDrive(0.01 + this.currentDrive * this.currentDrive / 500)
  }

  updateVolume (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    this.volNode.gain.value = value / 100
  }

  updateReverb (value) {
    if (value < 0 || value > 100) {
      throw new RangeError()
    }
    value = value / 100

    // equal-power crossfade
    var gain1 = Math.cos(value * 0.5 * Math.PI)
    var gain2 = Math.cos((1.0 - value) * 0.5 * Math.PI)

    this.revBypassGain.gain.value = gain1
    this.revGain.gain.value = gain2
  }

  changeModMultiplier () {
    this.modOscFreqMultiplier = (this.moDouble ? 2 : 1) * (this.moQuadruple ? 4 : 1)
    this.updateModFrequency(this.currentModFrequency)
  }

  onChangeOctave (ev) {
    this.currentOctave = ev.target.selectedIndex
  }

  initAudio () {
    // set up the master effects chain for all voices to connect to.
    this.effectChain = this.audioContext.createGain()
    this.waveshaper = new WaveShaper(this.audioContext)
    this.effectChain.connect(this.waveshaper.input)
    this.updateDrive(this.currentDrive)

    this.revNode = this.audioContext.createGain()
    this.revGain = this.audioContext.createGain()
    this.revBypassGain = this.audioContext.createGain()

    this.volNode = this.audioContext.createGain()
    this.volNode.gain.value = this.currentVol
    this.compressor = this.audioContext.createDynamicsCompressor()
    this.waveshaper.output.connect(this.revNode)
    this.waveshaper.output.connect(this.revBypassGain)
    this.revNode.connect(this.revGain)
    this.revGain.connect(this.volNode)
    this.revBypassGain.connect(this.volNode)
    this.updateReverb(this.currentRev)

    this.volNode.connect(this.compressor)
    this.compressor.connect(this.audioContext.destination)
    this.updateVolume(this.currentVol)

    const irRRequest = new XMLHttpRequest()
    irRRequest.open('GET', 'sounds/irRoom.wav', true)
    irRRequest.responseType = 'arraybuffer'
    irRRequest.onload = () => {
      this.audioContext.decodeAudioData(irRRequest.response, (buffer) => {
        this.revNode.buffer = buffer
      })
    }
    irRRequest.send()
  }

  destroy () {
    for (var i = 0; i < this.voices.length; i++) {
      delete this.voices[i]
    }
  }
}

module.exports = Synthesizer
